### DISCUSSION TITLE
How can I get the "root" directory inside a ScopeProvider?

### DISCUSSION THREAD

**Discussion by AignerGames** [eclipse-langium/langium#1775] [Q&A] [answered]
Hallo, I implement a custom ScopeProvider which uses the DefaultScopeProvider as a base and override the getGlobalScope with custom logic. I need a way to get the current "root" directory of the files (when they are edited inside VS Code), but I can't find any API to get this information. I tried to google and asked some chat bots, but they all gave me answers with an outdated API. 

I have access to the LangiumCoreServices, which has access to TextDocumentProvider, WorkspaceManager, FileSystemProvider, LangiumDocuments and many other things but I checked all of them and couldn't find any functions which would give me the root directory.

Additional info:
By default the scope provider returns all elements from all files, but I want to restrict this and only return elements from the current file and elements from files which are located in the directory root/Global, that's why I have to know the root directory to find the directory named "Global", as a workaround I currently only check the directory name, so the directory can be nested anywhere and there can also be multiple. It's only a hobby project so it doesn't matter too much and maybe it's even a good feature to allow multiple "Global" folders, but I was still interessted to know if there would be a way to get the root directory.

**Comment by cdietrich** [comment, marked as answer]
maybe calling `serivces.shared.workspace.WorkspaceManager.workspaceFolders()` does the trick

**Reply by msujew** [reply]
Yes, exactly. You should be able to use the [`WorkspaceManager#workspaceFolders`](https://eclipse-langium.github.io/langium/interfaces/langium.WorkspaceManager.html) getter together with the `uri` of the current document to find out what the *current* workspace folder is.

**Reply by AignerGames** [reply]
Thanks for the answer, it looks like I have a old version (3.2 instead of the current 3.3), because my version doesn't have the workspaceFolders. I will try it again once I manage to update the langium version. (Im new to TypeScript and VSCode node development and I get a timeout error everytime I try to update, but I will try it again after the holidays)

**Comment by joswarmer** [comment]
I am having the same requirement. The approach that I take is to call het getScope(...) from the DefaultScopeProvider and then filter it with only the items that come from a file in the same directory as where I an asking the scope for.  I can get the uri's of all elements involved, so this works fine in many cases, except for the following.

If I have two nodes with identical names in files in two different folders, the default scoper does not return both nodes, which is what I expected. I could then filter the one I need.  However, the default scoper only returns one of the identical named nodes, and if that is the one not from a file in the current folder, I won't see the one from the current folder.

Therefore it seems I need to dive deeper into the scoping calculation. I could either
-  get all nodes, including those with identical names, or
-  influence which nodes are selected if the are identical names to get the cporrect one.
 
Any help would be appreciated.

**Reply by msujew** [reply]
Hey @joswarmer,

by default, the `ScopeComputation` does not export **all** nodes in a document to the index, but only top level nodes. This is outlined in [our documentation on this topic](https://langium.org/docs/reference/document-lifecycle/#computing-scopes). You will need to adjust your `ScopeComputation` service to export the node to the global scope/index.

**Reply by joswarmer** [reply]
Hi @msujew , the nodes that I need are exported to the index, using my own ScopeComputation class. However, when two nodes with the same name are exported, I only see one back when calling the default scope provider. It seems that either two identically named nodes are not both stored in the index, or if they are both stored, the default scope provider only picks one of them.

**Reply by msujew** [reply]
Generally, both should be stored in the index, however, the default global scope `MapScope` will likely only pick one of that up (as it is a map). See also:

https://github.com/eclipse-langium/langium/blob/c3afb247d7dcfd41733c051b8db0395d50398b33/packages/langium/src/references/scope-provider.ts#L100-L102

**Reply by joswarmer** [reply]
So if I take the `this.indexManager.allElements(referenceType)` and filter out myself which node I need, I can then create the MapScope with the correct (for me) nodes. I'll try that out. 

Tnx, Jos

**Reply by msujew** [reply]
Ideally, you should probably adjust your `ScopeProvider` to create a local scope and a global scope, with the global scope being the parent of the local scope. That way the linker will prioritize linking to the local object instead of the one that is exported to the global scope.

**Reply by joswarmer** [reply]
In my case both nodes with the same name are not local, so this probably won't work.
